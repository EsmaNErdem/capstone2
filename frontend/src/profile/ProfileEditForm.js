import React, { useState } from 'react';
import BookClubApi from '../api';
import { Formik, Field, Form, ErrorMessage } from 'formik';
import { Typography, IconButton } from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import Alert from "../utilities/Alert"
import * as Yup from 'yup';
import "./FormContainer.css"

/**ProfileEditForm Component
 * Displays a form with and rendered controlled component functionality and uses modal functionality
 * 
 * * A form component for user profile update. Handles user input for, first name, last name, email and user image while username constant.
 * Displays form errors if registration fails.
 * On submission:
 * - calls API to user info
 * - update currentUser state
 * - shows conformation message when update is successful
 *
 * Routes ==> ProfileForm ==> Alert
 */
const ProfileEditForm = ({ close=false, closeModal, initialValues, updateUser }) => {
    console.debug("ProfileEditForm");

    const [error, setError] = useState(null);

    /**
     * Handles form submission. Calls the signup function with the form data.
     * Redirects to the home page upon successful registration, or displays form errors.
     */
    const handleSubmit = async (values, { setSubmitting }) => {
        try {
            const updatedUser =  await BookClubApi.updateUser(initialValues.username, values)
            updateUser(user => ({
                ...user,
                firstName: updatedUser.firstName,
                lastName: updatedUser.lastName,
                img: updatedUser.img,
                email: updatedUser.email,
            }))
            if (close) closeModal();
        } catch (error) {
            setError("An error occurred during edit.");
        } finally {
            setSubmitting(false);
        }
    };

    return (
        <div className='ProfileEditForm'>
            <Formik
                initialValues={{ firstName: initialValues.firstName, lastName: initialValues.lastName, email: initialValues.email, img: initialValues.img }}
                validationSchema={Yup.object({
                    password: Yup.string()
                        .notRequired(),
                    firstName: Yup.string()
                        .max(15, 'Must be 15 characters or less')
                        .notRequired(),
                    lastName: Yup.string()
                        .max(20, 'Must be 20 characters or less')
                        .notRequired(),
                    email: Yup.string()
                    .email('Invalid email address')
                    .notRequired(),
                    img: Yup.string().notRequired(),
                })}
                onSubmit={handleSubmit}
                >
                <Form className='FormContainer'>
                    <IconButton onClick={closeModal} sx={{ width: "20px"}}>
                        <CloseIcon />
                    </IconButton>
                    <Typography variant="h2" sx={{  marginBottom: 2 }}>Edit User</Typography>
                    {error ? <Alert type="danger" messages={[error]} />: null}
                    <div>
                        <label htmlFor="password">Password</label>
                        <Field id="password" name="password" type="password" placeholder="Password" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="password" /></span>
                    </div>

                    <div>
                        <label htmlFor="firstName">First Name</label>
                        <Field id="firstName" name="firstName" type="text" placeholder="First Name" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="firstName" /></span>
                    </div>

                    <div>
                        <label htmlFor="lastName">Last Name</label>
                        <Field id="lastName" name="lastName" type="text" placeholder="Last Name" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="lastName" /></span>
                    </div>

                    <div>
                        <label htmlFor="email">Email Address</label>
                        <Field id="email" name="email" type="email" placeholder="Email" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="email" /></span>
                    </div>

                    <div>
                        <label htmlFor="img">Profile Image URL</label>
                        <Field id="img" name="img" type="text" placeholder="Profile Image URL"/>
                        <span className='ErrorMessageContainer'><ErrorMessage name="img" /></span>
                    </div>

                    <button type="submit">Save Changes</button>
                </Form>
            </Formik>
        </div>
    );
};

export default ProfileEditForm;