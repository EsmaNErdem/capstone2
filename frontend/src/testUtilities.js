import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
