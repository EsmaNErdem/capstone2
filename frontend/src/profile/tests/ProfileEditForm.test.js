import { render, fireEvent, waitFor } from "@testing-library/react";
import ProfileEditForm from "../ProfileEditForm";
import { UserProvider } from "../../testUtilities";
import { MemoryRouter } from "react-router-dom";
import BookClubApi from '../../api';

const updateUser = jest.fn();

const initial = {
    username: "Test1",
    firstName: "TestFirs1",
    lastName: "TestLast1",
    img: "UserImg1",
    email: "u1@email.com",
}

const mockUpdatedUser = {
    firstName: "TestFirs",
    lastName: "TestLast",
    img: "UserImg",
    email: "u@email.com",
}

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        updateUser: jest.fn(),
    },
}));

test("renders without crashing", async () => {
    BookClubApi.updateUser.mockImplementationOnce(() => (mockUpdatedUser));

    render(
        <UserProvider>
            <MemoryRouter>
                <ProfileEditForm updateUser={updateUser} closeModal={()=>{}} initialValues={initial}/>
            </MemoryRouter>
        </UserProvider>
    ); 
});

test("it renders and matches with snaphot", () => {
    BookClubApi.updateUser.mockImplementationOnce(() => (mockUpdatedUser));

    const { asFragment } = render(
        <UserProvider>
            <MemoryRouter>
                <ProfileEditForm updateUser={updateUser} closeModal={()=>{}} initialValues={initial}/>
            </MemoryRouter>
        </UserProvider>
        );

    expect(asFragment()).toMatchSnapshot();
});

test("LoginForm renders correctly", () => {
    BookClubApi.updateUser.mockImplementationOnce(() => (mockUpdatedUser));

    const { getByLabelText, getByText } = render(
        <UserProvider>
            <MemoryRouter>
                <ProfileEditForm  updateUser={updateUser} closeModal={()=>{}} initialValues={initial}/>
            </MemoryRouter>
        </UserProvider>
    );
    
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email Address");
    const saveButton = getByText("Save Changes");
  
    expect(firstNameInput).toBeInTheDocument();
    expect(lastNameInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(saveButton).toBeInTheDocument();
  });
  
  test("Submitting the form calls login function", async () => {
    BookClubApi.updateUser.mockImplementationOnce(() => (mockUpdatedUser));

    const { getByLabelText, getByText } = render(
          <UserProvider >
            <MemoryRouter>
                  <ProfileEditForm updateUser={updateUser} closeModal={()=>{}} initialValues={initial}/>
            </MemoryRouter>
          </UserProvider>
      );
    
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email Address");
    const saveButton = getByText("Save Changes");
  
    fireEvent.change(firstNameInput, { target: { value: "user" } });
    fireEvent.change(lastNameInput, { target: { value: "name" } });
    fireEvent.change(emailInput, { target: { value: "user@email.com" } });
    fireEvent.click(saveButton);

    await waitFor(() => {
            expect(BookClubApi.updateUser).toHaveBeenCalledWith("Test1", 
            {
                firstName: "user",
                lastName: "name",
                email: "user@email.com",
                img: "UserImg1"
            });
        });
  });
   
