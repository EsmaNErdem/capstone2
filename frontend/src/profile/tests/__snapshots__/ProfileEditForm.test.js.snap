// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div
    class="ProfileEditForm"
  >
    <form
      action="#"
      class="FormContainer"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-agcxlm-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <h2
        class="MuiTypography-root MuiTypography-h2 css-vqgaav-MuiTypography-root"
      >
        Edit User
      </h2>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          id="firstName"
          name="firstName"
          placeholder="First Name"
          type="text"
          value="TestFirs1"
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          id="lastName"
          name="lastName"
          placeholder="Last Name"
          type="text"
          value="TestLast1"
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="email"
        >
          Email Address
        </label>
        <input
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          value="u1@email.com"
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="img"
        >
          Profile Image URL
        </label>
        <input
          id="img"
          name="img"
          placeholder="Profile Image URL"
          type="text"
          value="UserImg1"
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <button
        type="submit"
      >
        Save Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;
